### PR TITLE
Refactor `StormLaws`

### PR DESCRIPTION
**Pull request by ttim**
In this PR I refactored `StormLaws` class with a way to test same `Producer` on both `Storm` and `Memory` platforms and compare their results. In future this approach can be extended to `Scalding` platform as well with a very unified way of testing.

Also I've included test for https://github.com/twitter/summingbird/issues/725, where we build `Storm` topology which fails with `ClassCastException` in runtime.

### CODE REVIEW

**Comment by ttim** [comment]
@johnynek @pankajroark thoughts on this one?

**Comment by johnynek** [comment]
👍 

Nice cleanup.

**Comment by ttim** [comment]
@johnynek What I really like here is a fact we can introduce shared test suite for `Storm` and `Scalding` platforms after introduction of proper memory platform (with notion of time) and `Scalding` implementation for `CreatorCtx`.

Should I merge or better to wait for one more review from @pankajroark for example?

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala]: remove

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala]: Can you make this comment more clear, I'm not sure what it means in this context.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala]: Comment that this is not thread safe.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: When do we plan to remove these commented blocks.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: Can we create a helper method to simplify this code for readability:
testStormEqualToMemory(new ProducerCreator { def apply = fn })
to
testStormEqualToMemory(fn)

**Review by pankajroark — CHANGES_REQUESTED** [review]

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala
Comment (johnynek) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: I think we should make issues about this and see if they should be gone, or if we can find a way to make them pass. I haven't looked at them much. They may be assuming a stronger contract than actually exists.

Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: I've mentioned issue in code above (https://github.com/twitter/summingbird/issues/725), and it would be fixed soon when I'll land all my PRs. Also it's not about contract, it's just errors in our `Storm` platform implementation.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: I don't know how to do this because this method has type parameter. I changed impl thou, so now it looks even better.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala]: Done

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala]: Done

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala
Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormTestUtils.scala]: Done

**Comment by ttim** [comment]
@johnynek @pankajroark Hi, I've reworked this change a lot with introduction of `TestPlatform` which corresponds to previous `CreatorCtx`, can you take a look? I think it's much more Summingbirdish approach.

File: summingbird-core-test/src/main/scala/com/twitter/summingbird/TestPlatform.scala
Comment (pankajroark) [inline summingbird-core-test/src/main/scala/com/twitter/summingbird/TestPlatform.scala]: Can you comment here why ordering is needed for the Test Platform? As I understand it is because one of the other platforms requires it and without this we won't be able to transform. It's not very obvious though, so a comment would be useful.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala
Comment (pankajroark) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: In that case, IMO, we can just leave a note with the commented code for now.

**Review by pankajroark — APPROVED** [review]
lgtm

File: summingbird-core-test/src/main/scala/com/twitter/summingbird/TestPlatform.scala
Comment (ttim) [reply summingbird-core-test/src/main/scala/com/twitter/summingbird/TestPlatform.scala]: Removed `Ordering`. You are right - it's needed because concurrent platforms break ordering of elements. Add a comment where we check sinks content.

File: summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala
Comment (pankajroark) [inline summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: nit: This is also -> This also

Comment (ttim) [reply summingbird-storm-test/src/test/scala/com/twitter/summingbird/storm/StormLaws.scala]: Ha, fixed

**Pull request merged**
